File: java/benchmarks/src/main/java/glide/benchmarks/CompressionBenchmark.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.benchmarks;

import glide.api.GlideClient;
import glide.api.models.configuration.CompressionBackend;
import glide.api.models.configuration.CompressionConfiguration;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.NodeAddress;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.Random;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

/**
 * Comprehensive Java benchmark for GLIDE client compression performance.
 * 
 * This benchmark tests:
 * 1. Various realistic data types (JSON, logs, CSV, XML, etc.)
 * 2. Different compression levels and their impact on TPS
 * 3. Memory usage comparison between compressed and uncompressed
 * 4. Throughput measurements for different data sizes
 * 5. Compression effectiveness across different data patterns
 */
public class CompressionBenchmark {
    
    private static final String HOST = "localhost";
    private static final int PORT = 6379;
    private static final Random RANDOM = new Random(42); // Fixed seed for reproducible results
    
    // Benchmark configuration
    private static final int WARMUP_ITERATIONS = 100;
    private static final int BENCHMARK_ITERATIONS = 1000;
    private static final int TPS_TEST_DURATION_MS = 5000; // 5 seconds
    
    public static void main(String[] args) {
        System.out.println("🚀 GLIDE Java Compression Benchmark");
        System.out.println("=" + "=".repeat(60));
        System.out.println("Warmup iterations: " + WARMUP_ITERATIONS);
        System.out.println("Benchmark iterations: " + BENCHMARK_ITERATIONS);
        System.out.println("TPS test duration: " + TPS_TEST_DURATION_MS + "ms");
        System.out.println();
        
        try {
            // Load standardized test datasets
            Map<String, List<String>> datasets = loadTestDatasets();
            
            // Run comprehensive benchmarks
            benchmarkCompressionLevels(datasets);
            benchmarkDataTypes(datasets);
            benchmarkThroughput(datasets);
            benchmarkMemoryEfficiency(datasets);
            
            System.out.println("\n🎉 Benchmark completed successfully!");
            
        } catch (Exception e) {
            System.err.println("\n❌ Error during benchmark: " + e.getClass().getSimpleName() + ": " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Load standardized test datasets from files
     */
    private static Map<String, List<String>> loadTestDatasets() {
        System.out.println("📊 Loading standardized test datasets...");
        
        Map<String, List<String>> datasets = new HashMap<>();
        
        // Load datasets from standardized files
        datasets.put("json_objects", loadDatasetFromFile("json_objects.txt", false));
        datasets.put("app_logs", loadDatasetFromFile("app_logs.txt", false));
        datasets.put("csv_data", loadDatasetFromFile("csv_data.txt", false));
        datasets.put("xml_docs", loadDatasetFromFile("xml_docs.txt", true)); // XML uses --- separator
        datasets.put("base64", loadDatasetFromFile("base64_data.txt", false));
        datasets.put("repetitive", loadDatasetFromFile("repetitive_text.txt", false));
        datasets.put("random", loadDatasetFromFile("random_data.txt", false));
        datasets.put("mixed_web", loadDatasetFromFile("mixed_web.txt", true)); // Mixed web uses --- separator
        
        // Print dataset info
        for (Map.Entry<String, List<String>> entry : datasets.entrySet()) {
            List<String> values = entry.getValue();
            int totalSize = values.stream().mapToInt(String::length).sum();
            int avgSize = values.size() > 0 ? totalSize / values.size() : 0;
            System.out.printf("  %-12s: %3d entries, %,8d total bytes, %,5d avg bytes%n", 
                entry.getKey(), values.size(), totalSize, avgSize);
        }
        System.out.println();
        
        return datasets;
    }
    
    /**
     * Load dataset from a file
     */
    private static List<String> loadDatasetFromFile(String filename, boolean useSeparator) {
        List<String> entries = new ArrayList<>();
        String dataDir = "data/";
        
        try {
            java.nio.file.Path filePath = java.nio.file.Paths.get(dataDir + filename);
            if (!java.nio.file.Files.exists(filePath)) {
                throw new java.io.FileNotFoundException("File does not exist: " + filePath.toAbsolutePath());
            }
            String content = java.nio.file.Files.readString(filePath);
            
            if (useSeparator) {
                // Split by --- separator for XML and mixed web content
                String[] parts = content.split("---");
                for (String part : parts) {
                    String trimmed = part.trim();
                    if (!trimmed.isEmpty()) {
                        entries.add(trimmed);
                    }
                }
            } else {
                // Split by lines for other formats
                String[] lines = content.split("\n");
                for (String line : lines) {
                    String trimmed = line.trim();
                    if (!trimmed.isEmpty()) {
                        entries.add(trimmed);
                    }
                }
            }
        } catch (Exception e) {
            System.err.println("⚠️ Failed to load " + filename + ": " + e.getClass().getSimpleName() + " - " + e.getMessage());
            System.err.println("   Attempted path: " + dataDir + filename);
            System.err.println("   Using fallback generated data for this dataset");
            // Fallback to generated data if file loading fails
            return generateFallbackData(filename);
        }
        
        return entries;
    }
    
    /**
     * Generate fallback data if file loading fails
     */
    private static List<String> generateFallbackData(String filename) {
        List<String> fallback = new ArrayList<>();
        
        // Simple fallback data generation
        for (int i = 0; i < 10; i++) {
            fallback.add(String.format("Fallback data entry %d for %s", i, filename));
        }
        
        return fallback;
    }
    
    /**
     * Benchmark different compression levels
     */
    private static void benchmarkCompressionLevels(Map<String, List<String>> datasets) throws ExecutionException, InterruptedException {
        System.out.println("🎚️ Compression Level Benchmark");
        System.out.println("-".repeat(80));
        System.out.printf("%-6s %-12s %-10s %-10s %-8s %-8s%n", 
            "Level", "Dataset", "Original", "Compressed", "Ratio", "TPS");
        System.out.println("-".repeat(80));
        
        int[] levels = {1, 3, 6, 9, 15, 22};
        List<String> testDataset = datasets.get("json_objects"); // Use JSON for level testing
        
        for (int level : levels) {
            BenchmarkResult result = benchmarkConfiguration(
                createCompressionConfig(CompressionBackend.ZSTD, level, 64),
                "json_objects", testDataset
            );
            
            int originalSize = testDataset.stream().mapToInt(String::length).sum();
            System.out.printf("%-6d %-12s %,10d %,10d %-8.2f %,8.0f%n",
                level, "json_objects", 
                originalSize, result.compressedSize,
                result.compressionRatio, result.tps);
        }
        System.out.println();
    }
    
    /**
     * Benchmark different data types
     */
    private static void benchmarkDataTypes(Map<String, List<String>> datasets) throws ExecutionException, InterruptedException {
        System.out.println("📋 Data Type Compression Effectiveness");
        System.out.println("   Measures compression ratio and memory savings for different data types");
        System.out.println("-".repeat(80));
        System.out.printf("%-12s %-8s %-10s %-10s %-8s %-8s %-10s%n", 
            "Dataset", "Entries", "Original", "Redis Mem", "Ratio", "TPS", "Savings");
        System.out.println("-".repeat(80));
        
        CompressionConfiguration config = createCompressionConfig(CompressionBackend.ZSTD, 3, 64);
        
        for (Map.Entry<String, List<String>> entry : datasets.entrySet()) {
            String datasetName = entry.getKey();
            List<String> data = entry.getValue();
            
            BenchmarkResult result = benchmarkConfiguration(config, datasetName, data);
            int originalSize = data.stream().mapToInt(String::length).sum();
            double savings = ((double)(originalSize - result.compressedSize) / originalSize) * 100;
            
            System.out.printf("%-12s %,8d %,10d %,10d %-8.2f %,8.0f %9.1f%%%n",
                datasetName, data.size(), originalSize, result.compressedSize,
                result.compressionRatio, result.tps, savings);
        }
        System.out.println();
    }
    
    /**
     * Benchmark throughput with different configurations
     */
    private static void benchmarkThroughput(Map<String, List<String>> datasets) throws ExecutionException, InterruptedException {
        System.out.println("⚡ Throughput Benchmark (Operations/Second)");
        System.out.println("   Measures performance impact of compression on SET/GET operations");
        System.out.println("-".repeat(80));
        System.out.printf("%-12s %-12s %-12s %-12s%n", 
            "Dataset", "Uncompressed", "Compressed", "Perf Impact");
        System.out.println("-".repeat(80));
        
        CompressionConfiguration compressedConfig = createCompressionConfig(CompressionBackend.ZSTD, 3, 64);
        
        for (Map.Entry<String, List<String>> entry : datasets.entrySet()) {
            String datasetName = entry.getKey();
            List<String> data = entry.getValue();
            
            // Benchmark uncompressed
            BenchmarkResult uncompressed = benchmarkConfiguration(null, datasetName + "_uncomp", data);
            
            // Benchmark compressed
            BenchmarkResult compressed = benchmarkConfiguration(compressedConfig, datasetName + "_comp", data);
            
            double difference = ((compressed.tps - uncompressed.tps) / uncompressed.tps) * 100;
            

            
            System.out.printf("%-12s %,12.0f %,12.0f %+11.1f%%%n",
                datasetName, uncompressed.tps, compressed.tps, difference);
        }
        System.out.println();
    }
    
    /**
     * Benchmark memory efficiency
     */
    private static void benchmarkMemoryEfficiency(Map<String, List<String>> datasets) throws ExecutionException, InterruptedException {
        System.out.println("💾 Memory Efficiency Benchmark");
        System.out.println("-".repeat(80));
        System.out.printf("%-12s %-8s %-10s %-10s %-10s %-10s%n", 
            "Dataset", "Entries", "Original", "Uncomp Mem", "Comp Mem", "Mem Saved");
        System.out.println("-".repeat(80));
        
        CompressionConfiguration compressedConfig = createCompressionConfig(CompressionBackend.ZSTD, 3, 64);
        
        for (Map.Entry<String, List<String>> entry : datasets.entrySet()) {
            String datasetName = entry.getKey();
            List<String> data = entry.getValue();
            
            // Test uncompressed memory usage
            long uncompressedMemory = measureMemoryUsage(null, datasetName + "_mem_uncomp", data);
            
            // Test compressed memory usage
            long compressedMemory = measureMemoryUsage(compressedConfig, datasetName + "_mem_comp", data);
            
            int originalSize = data.stream().mapToInt(String::length).sum();
            double memorySaved = ((double)(uncompressedMemory - compressedMemory) / uncompressedMemory) * 100;
            
            System.out.printf("%-12s %,8d %,10d %,10d %,10d %9.1f%%%n",
                datasetName, data.size(), originalSize, uncompressedMemory, compressedMemory, memorySaved);
        }
        System.out.println();
    }
    
    /**
     * Benchmark a specific configuration
     */
    private static BenchmarkResult benchmarkConfiguration(CompressionConfiguration compressionConfig, 
                                                         String keyPrefix, List<String> dataList) 
            throws ExecutionException, InterruptedException {
        
        GlideClientConfiguration.GlideClientConfigurationBuilder configBuilder = 
            GlideClientConfiguration.builder()
                .address(NodeAddress.builder().host(HOST).port(PORT).build());
        
        if (compressionConfig != null) {
            configBuilder.compression(compressionConfig);
        }
        
        try (GlideClient client = GlideClient.createClient(configBuilder.build()).get()) {
            // Warmup - use different entries for each warmup iteration
            for (int i = 0; i < WARMUP_ITERATIONS; i++) {
                String data = dataList.get(i % dataList.size());
                client.set(keyPrefix + "_warmup_" + i, data).get();
            }
            
            // Measure TPS
            long startTime = System.currentTimeMillis();
            int operations = 0;
            
            while (System.currentTimeMillis() - startTime < TPS_TEST_DURATION_MS) {
                String data = dataList.get(operations % dataList.size());
                client.set(keyPrefix + "_tps_" + operations, data).get();
                client.get(keyPrefix + "_tps_" + operations).get();
                operations += 2; // SET + GET
            }
            
            long endTime = System.currentTimeMillis();
            double tps = (operations * 1000.0) / (endTime - startTime);
            
            // Measure compression across all entries
            long totalCompressedSize = 0;
            int originalSize = 0;
            
            for (int i = 0; i < dataList.size(); i++) {
                String data = dataList.get(i);
                String testKey = keyPrefix + "_size_test_" + i;
                client.set(testKey, data).get();
                long memUsage = getMemoryUsage(testKey, client);
                totalCompressedSize += memUsage;
                originalSize += data.length();
                

            }
            
            // Handle case where memory usage couldn't be measured
            if (totalCompressedSize == 0) {
                totalCompressedSize = originalSize; // Fallback to original size
            }
            
            // For uncompressed data, ratio should be 1.0
            double compressionRatio;
            if (compressionConfig == null) {
                compressionRatio = 1.0; // No compression applied
            } else {
                compressionRatio = totalCompressedSize > 0 ? (double) originalSize / totalCompressedSize : 1.0;
            }
            
            return new BenchmarkResult(tps, totalCompressedSize, compressionRatio);
        }
    }
    
    /**
     * Measure memory usage for a configuration
     */
    private static long measureMemoryUsage(CompressionConfiguration compressionConfig, 
                                          String keyPrefix, List<String> dataList) 
            throws ExecutionException, InterruptedException {
        
        GlideClientConfiguration.GlideClientConfigurationBuilder configBuilder = 
            GlideClientConfiguration.builder()
                .address(NodeAddress.builder().host(HOST).port(PORT).build());
        
        if (compressionConfig != null) {
            configBuilder.compression(compressionConfig);
        }
        
        try (GlideClient client = GlideClient.createClient(configBuilder.build()).get()) {
            long totalMemory = 0;
            for (int i = 0; i < dataList.size(); i++) {
                String key = keyPrefix + "_" + i;
                client.set(key, dataList.get(i)).get();
                totalMemory += getMemoryUsage(key, client);
            }
            return totalMemory;
        }
    }
    
    /**
     * Create compression configuration
     */
    private static CompressionConfiguration createCompressionConfig(CompressionBackend backend, 
                                                                   int level, 
                                                                   int minSize) {
        return CompressionConfiguration.builder()
                .enabled(true)
                .backend(backend)
                .compressionLevel(level)
                .minCompressionSize(minSize)
                .build();
    }
    
    /**
     * Get memory usage for a key
     */
    private static long getMemoryUsage(String key, GlideClient client) {
        try {
            Object result = client.customCommand(new String[]{"MEMORY", "USAGE", key}).get();
            if (result instanceof Long) {
                return (Long) result;
            } else if (result instanceof Integer) {
                return ((Integer) result).longValue();
            } else if (result != null) {
                // Try to parse as string number
                try {
                    return Long.parseLong(result.toString());
                } catch (NumberFormatException nfe) {
                    // Fall through to return 0
                }
            }
            return 0;
        } catch (Exception e) {
            // Silently return 0 for cleaner output
            return 0;
        }
    }
    

    
    /**
     * Benchmark result container
     */
    private static class BenchmarkResult {
        final double tps;
        final long compressedSize;
        final double compressionRatio;
        
        BenchmarkResult(double tps, long compressedSize, double compressionRatio) {
            this.tps = tps;
            this.compressedSize = compressedSize;
            this.compressionRatio = compressionRatio;
        }
    }
}

File: compression-docs/compression_benchmark.py
#!/usr/bin/env python3
"""
Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

Comprehensive Python benchmark for GLIDE client compression performance.

This benchmark tests:
1. Various realistic data types (JSON, logs, CSV, XML, etc.)
2. Different compression levels and their impact on TPS
3. Memory usage comparison between compressed and uncompressed
4. Throughput measurements for different data sizes
5. Compression effectiveness across different data patterns
"""

import asyncio
import json
import os
import statistics
import time
import sys
from pathlib import Path
from typing import Dict, List, Optional, Tuple

# Add the python directory to the path to import glide
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python', 'python'))

from glide import (
    GlideClient,
    GlideClientConfiguration,
    NodeAddress,
    CompressionConfiguration,
    CompressionBackend,
)


class CompressionBenchmark:
    def __init__(self, host: str = "localhost", port: int = 6379):
        self.host = host
        self.port = port
        self.warmup_iterations = 100
        self.benchmark_iterations = 1000
        self.tps_test_duration_ms = 5000  # 5 seconds
        
    def load_test_datasets(self) -> Dict[str, List[str]]:
        """Load standardized test datasets from files"""
        print("📊 Loading standardized test datasets...")
        
        datasets = {}
        data_dir = Path(__file__).parent.parent / "benchmarks" / "data"
        
        # Load datasets from standardized files
        datasets["json_objects"] = self._load_dataset_from_file(data_dir / "json_objects.txt", False)
        datasets["app_logs"] = self._load_dataset_from_file(data_dir / "app_logs.txt", False)
        datasets["csv_data"] = self._load_dataset_from_file(data_dir / "csv_data.txt", False)
        datasets["xml_docs"] = self._load_dataset_from_file(data_dir / "xml_docs.txt", True)  # XML uses --- separator
        datasets["base64"] = self._load_dataset_from_file(data_dir / "base64_data.txt", False)
        datasets["repetitive"] = self._load_dataset_from_file(data_dir / "repetitive_text.txt", False)
        datasets["random"] = self._load_dataset_from_file(data_dir / "random_data.txt", False)
        datasets["mixed_web"] = self._load_dataset_from_file(data_dir / "mixed_web.txt", True)  # Mixed web uses --- separator
        
        # Print dataset info
        for name, values in datasets.items():
            total_size = sum(len(value) for value in values)
            avg_size = total_size // len(values) if values else 0
            print(f"  {name:<12}: {len(values):3d} entries, {total_size:8,d} total bytes, {avg_size:5,d} avg bytes")
        
        print()
        return datasets
    
    def _load_dataset_from_file(self, filename: Path, use_separator: bool) -> List[str]:
        """Load dataset from a file"""
        try:
            with open(filename, 'r', encoding='utf-8') as f:
                content = f.read()
            
            if use_separator:
                # Split by --- separator for XML and mixed web content
                parts = content.split('---')
                entries = [part.strip() for part in parts if part.strip()]
            else:
                # Split by lines for other formats
                lines = content.split('\n')
                entries = [line.strip() for line in lines if line.strip()]
            
            return entries
            
        except Exception as e:
            print(f"⚠️ Failed to load {filename}: {type(e).__name__} - {e}")
            print(f"   Using fallback generated data for this dataset")
            return self._generate_fallback_data(filename.name)
    
    def _generate_fallback_data(self, filename: str) -> List[str]:
        """Generate fallback data if file loading fails"""
        return [f"Fallback data entry {i} for {filename}" for i in range(10)]
    
    async def benchmark_compression_levels(self, datasets: Dict[str, List[str]]):
        """Benchmark different compression levels"""
        print("🎚️ Compression Level Benchmark")
        print("-" * 80)
        print(f"{'Level':<6} {'Dataset':<12} {'Original':<10} {'Compressed':<10} {'Ratio':<8} {'TPS':<8}")
        print("-" * 80)
        
        levels = [1, 3, 6, 9, 15, 22]
        test_dataset = datasets["json_objects"]  # Use JSON for level testing
        
        for level in levels:
            result = await self._benchmark_configuration(
                self._create_compression_config(CompressionBackend.ZSTD, level, 64),
                "json_objects", test_dataset
            )
            
            original_size = sum(len(data) for data in test_dataset)
            print(f"{level:<6d} {'json_objects':<12} {original_size:10,d} {result.compressed_size:10,d} {result.compression_ratio:<8.2f} {result.tps:8,.0f}")
        
        print()
    
    async def benchmark_data_types(self, datasets: Dict[str, List[str]]):
        """Benchmark different data types"""
        print("📋 Data Type Compression Effectiveness")
        print("   Measures compression ratio and memory savings for different data types")
        print("-" * 80)
        print(f"{'Dataset':<12} {'Entries':<8} {'Original':<10} {'Redis Mem':<10} {'Ratio':<8} {'TPS':<8} {'Savings':<10}")
        print("-" * 80)
        
        config = self._create_compression_config(CompressionBackend.ZSTD, 1, 64)
        
        for dataset_name, data in datasets.items():
            result = await self._benchmark_configuration(config, dataset_name, data)
            original_size = sum(len(entry) for entry in data)
            savings = ((original_size - result.compressed_size) / original_size) * 100
            
            print(f"{dataset_name:<12} {len(data):8,d} {original_size:10,d} {result.compressed_size:10,d} {result.compression_ratio:<8.2f} {result.tps:8,.0f} {savings:9.1f}%")
        
        print()
    
    async def benchmark_throughput(self, datasets: Dict[str, List[str]]):
        """Benchmark throughput with different configurations"""
        print("⚡ Throughput Benchmark (Operations/Second)")
        print("   Measures performance impact of compression on SET/GET operations")
        print("-" * 80)
        print(f"{'Dataset':<12} {'Uncompressed':<12} {'Compressed':<12} {'Perf Impact':<12}")
        print("-" * 80)
        
        compressed_config = self._create_compression_config(CompressionBackend.ZSTD, 1, 64)
        
        for dataset_name, data in datasets.items():
            # Benchmark uncompressed
            uncompressed = await self._benchmark_configuration(None, f"{dataset_name}_uncomp", data)
            
            # Benchmark compressed
            compressed = await self._benchmark_configuration(compressed_config, f"{dataset_name}_comp", data)
            
            difference = ((compressed.tps - uncompressed.tps) / uncompressed.tps) * 100
            
            print(f"{dataset_name:<12} {uncompressed.tps:12,.0f} {compressed.tps:12,.0f} {difference:+11.1f}%")
        
        print()
    
    async def benchmark_memory_efficiency(self, datasets: Dict[str, List[str]]):
        """Benchmark memory efficiency"""
        print("💾 Memory Efficiency Benchmark")
        print("-" * 80)
        print(f"{'Dataset':<12} {'Entries':<8} {'Original':<10} {'Uncomp Mem':<10} {'Comp Mem':<10} {'Mem Saved':<10}")
        print("-" * 80)
        
        compressed_config = self._create_compression_config(CompressionBackend.ZSTD, 1, 64)
        
        for dataset_name, data in datasets.items():
            # Test uncompressed memory usage
            uncompressed_memory = await self._measure_memory_usage(None, f"{dataset_name}_mem_uncomp", data)
            
            # Test compressed memory usage
            compressed_memory = await self._measure_memory_usage(compressed_config, f"{dataset_name}_mem_comp", data)
            
            original_size = sum(len(entry) for entry in data)
            memory_saved = ((uncompressed_memory - compressed_memory) / uncompressed_memory) * 100 if uncompressed_memory > 0 else 0
            
            print(f"{dataset_name:<12} {len(data):8,d} {original_size:10,d} {uncompressed_memory:10,d} {compressed_memory:10,d} {memory_saved:9.1f}%")
        
        print()
    
    async def _benchmark_configuration(self, compression_config: Optional[CompressionConfiguration], 
                                     key_prefix: str, data_list: List[str]) -> 'BenchmarkResult':
        """Benchmark a specific configuration"""
        
        config_builder = GlideClientConfiguration(
            addresses=[NodeAddress(host=self.host, port=self.port)]
        )
        
        if compression_config:
            config_builder.compression = compression_config
        
        client = await GlideClient.create(config_builder)
        
        try:
            # Warmup - use different entries for each warmup iteration
            for i in range(self.warmup_iterations):
                data = data_list[i % len(data_list)]
                await client.set(f"{key_prefix}_warmup_{i}", data)
            
            # Measure TPS
            start_time = time.time()
            operations = 0
            
            while (time.time() - start_time) * 1000 < self.tps_test_duration_ms:
                data = data_list[operations % len(data_list)]
                await client.set(f"{key_prefix}_tps_{operations}", data)
                await client.get(f"{key_prefix}_tps_{operations}")
                operations += 2  # SET + GET
            
            end_time = time.time()
            tps = operations / (end_time - start_time)
            
            # Measure compression across all entries
            total_compressed_size = 0
            original_size = 0
            
            for i, data in enumerate(data_list):
                test_key = f"{key_prefix}_size_test_{i}"
                await client.set(test_key, data)
                memory_usage = await self._get_memory_usage(test_key, client)
                total_compressed_size += memory_usage
                original_size += len(data)
            
            # Handle case where memory usage couldn't be measured
            if total_compressed_size == 0:
                total_compressed_size = original_size  # Fallback to original size
            
            # For uncompressed data, ratio should be 1.0
            if compression_config is None:
                compression_ratio = 1.0  # No compression applied
            else:
                compression_ratio = original_size / total_compressed_size if total_compressed_size > 0 else 1.0
            
            return BenchmarkResult(tps, total_compressed_size, compression_ratio)
            
        finally:
            await client.close()
    
    async def _measure_memory_usage(self, compression_config: Optional[CompressionConfiguration], 
                                   key_prefix: str, data_list: List[str]) -> int:
        """Measure memory usage for a configuration"""
        
        config_builder = GlideClientConfiguration(
            addresses=[NodeAddress(host=self.host, port=self.port)]
        )
        
        if compression_config:
            config_builder.compression = compression_config
        
        client = await GlideClient.create(config_builder)
        
        try:
            total_memory = 0
            for i, data in enumerate(data_list):
                key = f"{key_prefix}_{i}"
                await client.set(key, data)
                total_memory += await self._get_memory_usage(key, client)
            return total_memory
            
        finally:
            await client.close()
    
    def _create_compression_config(self, backend: CompressionBackend, level: int, min_size: int) -> CompressionConfiguration:
        """Create compression configuration"""
        return CompressionConfiguration(
            enabled=True,
            backend=backend,
            compression_level=level,
            min_compression_size=min_size
        )
    
    async def _get_memory_usage(self, key: str, client: GlideClient) -> int:
        """Get memory usage for a key"""
        try:
            result = await client.custom_command(["MEMORY", "USAGE", key])
            return int(result) if result is not None else 0
        except Exception:
            return 0


class BenchmarkResult:
    def __init__(self, tps: float, compressed_size: int, compression_ratio: float):
        self.tps = tps
        self.compressed_size = compressed_size
        self.compression_ratio = compression_ratio


async def main():
    print("🚀 GLIDE Python Compression Benchmark")
    print("=" * 61)
    print(f"Warmup iterations: 100")
    print(f"Benchmark iterations: 1000")
    print(f"TPS test duration: 5000ms")
    print()
    
    try:
        benchmark = CompressionBenchmark()
        
        # Load standardized test datasets
        datasets = benchmark.load_test_datasets()
        
        # Run comprehensive benchmarks
        await benchmark.benchmark_compression_levels(datasets)
        await benchmark.benchmark_data_types(datasets)
        await benchmark.benchmark_throughput(datasets)
        await benchmark.benchmark_memory_efficiency(datasets)
        
        print("\n🎉 Benchmark completed successfully!")
        
    except Exception as e:
        print(f"\n❌ Error during benchmark: {type(e).__name__}: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())
